File: leetcode/0029-divide-two-integers.js
var divide = function(dividend, divisor) {
    // 오버플로우 예외 처리
    if (dividend === -2147483648 && divisor === -1) {
        return 2147483647;
    }

    let multiplier = 1;

    if (dividend < 0) {
        dividend = -dividend;
        multiplier = -multiplier;
    }

    if (divisor < 0) {
        divisor = -divisor;
        multiplier = -multiplier;
    }

    let left = dividend;
    let quotient = 0;

    while (left >= divisor) {
        let temporalQuotient = 1;
        let currentNumber = divisor;

        while (true) {
            const nextNumber = currentNumber * 2;

            if (nextNumber <= left) {
                temporalQuotient *= 2;
                currentNumber = nextNumber;
            } else {
                left -= currentNumber;
                quotient += temporalQuotient;
                break;
            }
        }
    }

    return quotient * multiplier;
};

File: leetcode/0029-divide-two-integers.java
class Solution {
    public int divide(int dividend, int divisor) {
        // 오버플로우 예외 처리
        if (dividend == Integer.MIN_VALUE && divisor == -1) {
            return Integer.MAX_VALUE;
        }

        int multiplier = 1;

        // Integer.MIN_VALUE의 절댓값을 처리하기 위해 long 사용
        long dividendLong = dividend;
        long divisorLong = divisor;

        if (dividendLong < 0) {
            dividendLong = -dividendLong;
            multiplier = -multiplier;
        }

        if (divisorLong < 0) {
            divisorLong = -divisorLong;
            multiplier = -multiplier;
        }

        long left = dividendLong;
        long quotient = 0;

        while (left >= divisorLong) {
            long temporalQuotient = 1;
            long currentNumber = divisorLong;

            while (true) {
                long nextNumber = currentNumber << 1;

                if (nextNumber <= left) {
                    temporalQuotient <<= 1;
                    currentNumber = nextNumber;
                } else {
                    left -= currentNumber;
                    quotient += temporalQuotient;
                    break;
                }
            }
        }

        return (int) (quotient * multiplier);
    }
}